lib/elixir_sense/core/metadata_builder.ex

```diff
@@ -1663,6 +1668,98 @@ defmodule ElixirSense.Core.MetadataBuilder do
     {ast, {vars, match_context}}
   end
 
+  def infer_type_from_guards(guard_ast, vars, state) do
+    type_info = type_information_from_guards(guard_ast, state)
+
+    Enum.reduce(type_info, vars, fn {var, type}, acc ->
+      index = Enum.find_index(acc, &(&1.name == var))
+
+      if index,
+        do: List.update_at(acc, index, &Map.put(&1, :type, type)),
+        else: acc
+    end)
+  end
+
+  # A guard expression can be in either these form:
+  #        :and                          :or
+  #      /     \            or          /   \              or      guard_expr
+  # guard_expr  guard_expr        guard_expr  guard_expr
+  #
+  # type information from :and subtrees are mergeable
+  # type information from :or subtrees are discarded
+  defp type_information_from_guards({:and, _, [guard_l, guard_r]}, state) do
```

Comment:
Please don't do evaluation here and leave it for `Binding` module
```suggestion
  defp type_information_from_guards({:and, _, [guard_l, guard_r]}, state), do: [{:intersection, [guard_l, guard_r]}]
```

Reply:
I'm not sure if the current implementation in binding will merge map keys though

Reply:
Fixed in e26604161873790c709ec1e9524e1cfd7d0911d4. 

> I'm not sure if the current implementation in binding will merge map keys though

Based on [this test case](https://github.com/elixir-lsp/elixir_sense/blob/master/test/elixir_sense/core/binding_test.exs#L1911-L1934), I think we can, right?

Reply:
You're right @Goose97